### PR TITLE
feat: add DNS tunnel support

### PR DESCRIPTION
**Pull request by garmr-ulfr**
This pull request introduces support for DNS tunneling (DNSTT) by adding a new configuration structure, updating the HTTP client initialization, and modifying the configuration generation process. Key changes include the integration of DNSTT configuration, updates to the `kindling` library, and enhancements to the configuration generator and tests.

### Integration of DNSTT Configuration:
* Added a new `DNSTTConfig` struct in `common/httpclient.go` to store DNS tunneling settings, including domain, public key, and DNS resolver details. A default configuration is initialized in the `init` function. (`[common/httpclient.goL39-R77](diffhunk://#diff-485d80568d05d3233923ec91d33f7b5f7dc69b223d233809b6f8a794589494d1L39-R77)`)
* Introduced `SetDNSTTConfig` and `newDNSTT` functions in `common/httpclient.go` to update and create DNSTT instances. These are integrated into the `GetHTTPClient` function to enable DNS tunneling. (`[[1]](diffhunk://#diff-485d80568d05d3233923ec91d33f7b5f7dc69b223d233809b6f8a794589494d1L39-R77)`, `[[2]](diffhunk://#diff-485d80568d05d3233923ec91d33f7b5f7dc69b223d233809b6f8a794589494d1R87-R105)`, `[[3]](diffhunk://#diff-485d80568d05d3233923ec91d33f7b5f7dc69b223d233809b6f8a794589494d1R143-R169)`)

### Configuration Generator Updates:
* Modified the `genconfig` tool to support reading DNSTT configuration from a YAML file via a new `--dnstt-file` flag. The configuration is validated and incorporated into the generated output. (`[[1]](diffhunk://#diff-e082ee5af8475ee1c113b5e3b1584eaf1585370280dde9c4d30f2e92289b736bR53)`, `[[2]](diffhunk://#diff-e082ee5af8475ee1c113b5e3b1584eaf1585370280dde9c4d30f2e92289b736bR259-R275)`, `[[3]](diffhunk://#diff-e082ee5af8475ee1c113b5e3b1584eaf1585370280dde9c4d30f2e92289b736bR373-R399)`)
* Updated the `GenerateConfig` method to include DNSTT configuration in the global configuration model. (`[[1]](diffhunk://#diff-e082ee5af8475ee1c113b5e3b1584eaf1585370280dde9c4d30f2e92289b736bL201-R214)`, `[[2]](diffhunk://#diff-e082ee5af8475ee1c113b5e3b1584eaf1585370280dde9c4d30f2e92289b736bR664)`)

### Configuration and Tests:
* Added DNSTT configuration placeholders to `embeddedconfig/global.yaml.tmpl` and `genconfig/testdata/global_test.yaml.tmpl`. (`[[1]](diffhunk://#diff-2a3f1cda1c632f796879e40b01b2b015af61a1f2f4dac9365b4df68489509b99R431-R437)`, `[[2]](diffhunk://#diff-2609c921bfab6d4072afadb28e43a78234b2a94e12dc92a2a6bfe3b560ecb5deR422-R428)`)
* Enhanced tests in `genconfig_test.go` to verify that DNSTT configuration is correctly integrated into the generated configuration. (`[[1]](diffhunk://#diff-b5380bcac2e88c8dceeaa667327d5b5f24cfa94f86375d30fb9f978ca4accd64R50-R57)`, `[[2]](diffhunk://#diff-b5380bcac2e88c8dceeaa667327d5b5f24cfa94f86375d30fb9f978ca4accd64R99-R101)`, `[[3]](diffhunk://#diff-b5380bcac2e88c8dceeaa667327d5b5f24cfa94f86375d30fb9f978ca4accd64L164-R182)`)

### Dependency Updates:
* Updated the `kindling` library version in `go.mod` to include support for DNS tunneling. (`[go.modL197-R197](diffhunk://#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6L197-R197)`)

closes getlantern/engineering/issues/2378

### CODE REVIEW

File: common/httpclient.go [hunk 1]
```diff
@@ -53,26 +84,25 @@ func GetHTTPClient() *http.Client {
...
-			kindling.WithDomainFronting(f),
-			kindling.WithProxyless(domains...),
-		)
+		kOptions = append(kOptions, kindling.WithDomainFronting(f))
 	}
+	if d, err := newDNSTT(); err != nil {
+		log.Errorf("Failed to create DNSTT: %v", err)
+	} else {
+		kOptions = append(kOptions, kindling.WithDNSTunnel(d))
+	}
```
[the remark this answers is not in the document]
Comment (garmr-ulfr) [reply common/httpclient.go]: > but the code structure makes this hard to follow

It's really not..

File: genconfig/genconfig_test.go [hunk 2]
```diff
@@ -56,6 +66,7 @@ func TestGenerateConfig(t *testing.T) {
 		givenMinFrequency    float64
 		givenMinMasquerades  int
 		givenMaxMasquerades  int
+		givenDNSTTConfig     *common.DNSTTConfig
```
Comment (WendelHime) [inline genconfig/genconfig_test.go:69]: Could you please add a test scenario where the given dnstt config is nil? If we merge the PR right now, the default value for the update_masquerade will be nil and if it breaks, it'll stop updating the global config

**Review by WendelHime — APPROVED** [review]
In general it LGTM, just remember the genconfig changes need to be added to lantern-cloud update_masquerade too in order to propagate the configuration. I've left one comment suggestion

File: genconfig/genconfig_test.go (hunk 2 above)
Comment (WendelHime) [reply genconfig/genconfig_test.go:69]: If this is a optional configuration, you might need to add something like

```gotmpl
{{if .dnstt}}
dnsttconfig:
  domain: "{{.dnstt.Domain}}"
  publicKey: "{{.dnstt.PublicKey}}"
  dohResolver: "{{.dnstt.DoHResolver}}"
  dotResolver: "{{.dnstt.DoTResolver}}"
  utlsDistribution: "{{.dnstt.UTLSDistribution}}"
{{end}}
```

Comment (garmr-ulfr) [reply genconfig/genconfig_test.go:69]: Well, I did that and found an issue where `global.DNSTTConfig` was always being set somewhere..I could not find where..but I fixed it and added nil check test.

**Comment by garmr-ulfr** [comment]
> just remember the genconfig changes need to be added to lantern-cloud update_masquerade too in order to propagate the configuration.

I did not know that. Thanks!

**Comment by garmr-ulfr** [comment]
@WendelHime would it break anything if we were to merge it as is now?

**Comment by WendelHime** [comment]
> @WendelHime would it break anything if we were to merge it as is now?

I think is good to go. If the parameter is acceptable as nullable, I'm not expecting conflicts with older versions
